pkg/types/tuf/v0.0.1/entry.go

```diff
@@ -0,0 +1,406 @@
+
+		signature, err := artifactFactory.NewSignature(sigReadCloser)
+		if err != nil {
+			return closePipesOnError(types.ValidationError(err))
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case sigResult <- signature:
+			return nil
+		}
+	})
+
+	keyResult := make(chan pki.PublicKey)
+
+	g.Go(func() error {
+		defer close(keyResult)
+
+		keyReadCloser, err := util.FileOrURLReadCloser(ctx, v.TufObj.Root.Signed.URL.String(),
+			v.TufObj.Root.Signed.Content)
+		if err != nil {
+			return closePipesOnError(err)
+		}
+		defer keyReadCloser.Close()
+
+		key, err := artifactFactory.NewPublicKey(keyReadCloser)
+		if err != nil {
+			return closePipesOnError(types.ValidationError(err))
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case keyResult <- key:
+			return nil
+		}
+	})
+
+	// the sigObj contains the signed content.
+	g.Go(func() error {
+		v.keyObj, v.sigObj = <-keyResult, <-sigResult
+
+		if v.keyObj == nil || v.sigObj == nil {
+			return closePipesOnError(errors.New("failed to read signature or public key"))
+		}
+
+		var err error
+		if err = v.sigObj.Verify(nil, v.keyObj); err != nil {
+			return closePipesOnError(types.ValidationError(err))
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			return nil
+		}
+	})
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	v.fetchedExternalEntities = true
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if err := v.FetchExternalEntities(ctx); err != nil {
+		return nil, err
+	}
+	if v.sigObj == nil {
+		return nil, errors.New("signature object not initialized before canonicalization")
+	}
+	if v.keyObj == nil {
+		return nil, errors.New("key object not initialized before canonicalization")
+	}
+
+	canonicalEntry := models.TufV001Schema{}
+	canonicalEntry.ExtraData = v.TufObj.ExtraData
+
+	// need to canonicalize manifest (canonicalize JSON)
+	var err error
+	canonicalEntry.Root = &models.TufManifestV001Schema{Signed: &models.TufManifestV001SchemaSigned{}}
+	canonicalEntry.Root.Signed.Content, err = v.keyObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+	// fill in the rest of the manifest fields from the content
+	canonicalEntry.Root, err = updateTufManifest(canonicalEntry.Root)
+	if err != nil {
+		return nil, err
+	}
+
+	canonicalEntry.Manifest = &models.TufManifestV001Schema{Signed: &models.TufManifestV001SchemaSigned{}}
+	canonicalEntry.Manifest.Signed.Content, err = v.sigObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+	canonicalEntry.Manifest, err = updateTufManifest(canonicalEntry.Manifest)
+	if err != nil {
+		return nil, err
+	}
+	// wrap in valid object with kind and apiVersion set
+	tuf := models.Tuf{}
+	tuf.APIVersion = swag.String(APIVERSION)
+	tuf.Spec = &canonicalEntry
+
+	return json.Marshal(&tuf)
+}
+
+// Validate performs cross-field validation for fields in object
+// FIXME: we can probably export ValidateMetablock on in-toto.go
```

Comment:
Is this comment intended? Seems like a leftover.

Reply:
This was a Santiago comment -- not sure what it meant @SantiagoTorres if it's important let me know!

Reply:
Ah, this used to be my attempt at validating signature/structure of the wrapper back in the day. I think we did [export the method](https://github.com/in-toto/in-toto-golang/blob/f50646681592db791f6708f58ef8ec974f5b480d/in_toto/model.go#L822-L825) but this is probably handled by the proto stuff now :)